File: openmrs-client/src/main/java/org/openmrs/mobile/databases/AppDatabaseHelper.java
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */

package org.openmrs.mobile.databases;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import org.openmrs.mobile.application.OpenMRS;
import org.openmrs.mobile.dao.EncounterDAO;
import org.openmrs.mobile.dao.ObservationDAO;
import org.openmrs.mobile.dao.PatientDAO;
import org.openmrs.mobile.databases.entities.AllergyEntity;
import org.openmrs.mobile.databases.entities.ConceptEntity;
import org.openmrs.mobile.databases.entities.EncounterEntity;
import org.openmrs.mobile.databases.entities.LocationEntity;
import org.openmrs.mobile.databases.entities.ObservationEntity;
import org.openmrs.mobile.databases.entities.PatientEntity;
import org.openmrs.mobile.databases.entities.VisitEntity;
import org.openmrs.mobile.models.Allergen;
import org.openmrs.mobile.models.Allergy;
import org.openmrs.mobile.models.Encounter;
import org.openmrs.mobile.models.EncounterType;
import org.openmrs.mobile.models.Observation;
import org.openmrs.mobile.models.Patient;
import org.openmrs.mobile.models.PatientIdentifier;
import org.openmrs.mobile.models.PersonAddress;
import org.openmrs.mobile.models.PersonName;
import org.openmrs.mobile.models.Resource;
import org.openmrs.mobile.models.Visit;
import org.openmrs.mobile.models.VisitType;
import org.openmrs.mobile.utilities.ApplicationConstants;
import org.openmrs.mobile.utilities.DateUtils;
import org.openmrs.mobile.utilities.FormService;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

import rx.Observable;
import rx.schedulers.Schedulers;

public class AppDatabaseHelper {

    public static ObservationEntity observationToEntity(Observation obs, long encounterID) {
        ObservationEntity observationEntity = new ObservationEntity();
        observationEntity.setId(obs.getId());
        observationEntity.setUuid(obs.getUuid());
        observationEntity.setDisplay(obs.getDisplay());
        observationEntity.setEncounterKeyID(encounterID);
        observationEntity.setDisplayValue(obs.getDisplayValue());
        observationEntity.setDiagnosisOrder(obs.getDiagnosisOrder());
        observationEntity.setDiagnosisList(obs.getDiagnosisList());
        observationEntity.setDiagnosisCertainty(obs.getDiagnosisCertainty());
        observationEntity.setDiagnosisNote(obs.getDiagnosisNote());
        if (obs.getConcept() != null) {
            observationEntity.setConceptuuid(obs.getConcept().getUuid());
        } else {
            observationEntity.setConceptuuid(null);
        }
        return observationEntity;
    }

    public static List<Observation> observationEntityToObservation(List<ObservationEntity> observationEntityList) {
        List<Observation> observationList = new ArrayList<>();
        for (ObservationEntity entity : observationEntityList) {
            Observation obs = new Observation();
            obs.setId(entity.getId());
            obs.setEncounterID(entity.getEncounterKeyID());
            obs.setUuid(entity.getUuid());
            obs.setDisplay(entity.getDisplay());
            obs.setDisplayValue(entity.getDisplayValue());
            obs.setDiagnosisOrder(entity.getDiagnosisOrder());
            obs.setDiagnosisList(entity.getDiagnosisList());
            obs.setDiagnosisCertanity(entity.getDiagnosisCertainty());
            obs.setDiagnosisNote(entity.getDiagnosisNote());
            ConceptEntity concept = new ConceptEntity();
            concept.setUuid(entity.getConceptuuid());
            obs.setConcept(concept);
            observationList.add(obs);
        }
        return observationList;
    }

    public static EncounterEntity encounterToEntity(Encounter encounter, Long visitID) {
        EncounterEntity encounterEntity = new EncounterEntity();
        encounterEntity.setId(encounter.getId());
        encounterEntity.setDisplay(encounter.getDisplay());
        encounterEntity.setUuid(encounter.getUuid());
        if (visitID != null) {
            encounterEntity.setVisitKeyId(visitID.toString());
        }
        encounterEntity.setEncounterDateTime(encounter.getEncounterDatetime().toString());
        encounterEntity.setEncounterType(encounter.getEncounterType().getDisplay());
        encounterEntity.setPatientUuid(encounter.getPatientUUID());
        encounterEntity.setFormUuid(encounter.getFormUuid());
        if (null == encounter.getLocation()) {
            encounterEntity.setLocationUuid(null);
        } else {
            encounterEntity.setLocationUuid(encounter.getLocation().getUuid());
        }

        if (0 == encounter.getEncounterProviders().size()) {
            encounterEntity.setEncounterProviderUuid(null);
        } else {
            encounterEntity.setEncounterProviderUuid(encounter.getEncounterProviders().get(0).getUuid());
        }

        return encounterEntity;
    }

    public static Encounter encounterEntityToEncounter(EncounterEntity entity) {
        Encounter encounter = new Encounter();
        if (null != entity.getEncounterType()) {
            encounter.setEncounterType(new EncounterType(entity.getEncounterType()));
        }
        encounter.setId(entity.getId());
        if (null != entity.getVisitKeyId()) {
            encounter.setVisitID(Long.parseLong(entity.getVisitKeyId()));
        }
        encounter.setUuid(entity.getUuid());
        encounter.setDisplay(entity.getDisplay());
        Long dateTime = Long.parseLong(entity.getEncounterDateTime());
        encounter.setEncounterDatetime(DateUtils.convertTime(dateTime, DateUtils.OPEN_MRS_REQUEST_FORMAT));
        encounter.setObservations(new ObservationDAO().findObservationByEncounterID(entity.getId()));
        LocationEntity location;
        try {
            location = AppDatabase
                    .getDatabase(OpenMRS.getInstance().getApplicationContext())
                    .locationRoomDAO()
                    .findLocationByUUID(entity.getLocationUuid())
                    .blockingGet();
        } catch (Exception e) {
            location = null;
        }
        encounter.setLocation(location);
        encounter.setForm(FormService.getFormByUuid(entity.getFormUuid()));
        return encounter;
    }

    public static Visit visitEntityToVisit(VisitEntity visitEntity) {
        Visit visit = new Visit();
        visit.setId(visitEntity.getId());
        visit.setUuid(visitEntity.getUuid());
        visit.setDisplay(visitEntity.getDisplay());
        visit.setVisitType(new VisitType(visitEntity.getVisitType()));
        try {
            LocationEntity locationEntity = AppDatabase
                    .getDatabase(OpenMRS.getInstance().getApplicationContext())
                    .locationRoomDAO()
                    .findLocationByName(visitEntity.getVisitPlace())
                    .blockingGet();
            visit.setLocation(locationEntity);
        } catch (Exception e) {
            visit.setLocation(new LocationEntity(visitEntity.getVisitPlace()));
        }
        visit.setStartDatetime(visitEntity.getStartDate());
        visit.setStopDatetime(visitEntity.getStopDate());
        visit.setEncounters(new EncounterDAO().findEncountersByVisitID(visitEntity.getId()));
        visit.setPatient(new PatientDAO().findPatientByID(String.valueOf(visitEntity.getPatientKeyID())));

        return visit;
    }

    public static VisitEntity visitToVisitEntity(Visit visit) {
        VisitEntity visitEntity = new VisitEntity();
        visitEntity.setId(visit.getId());
        visitEntity.setUuid(visit.getUuid());
        visitEntity.setPatientKeyID(visit.getPatient().getId());
        visitEntity.setVisitType(visit.getVisitType().getDisplay());
        visitEntity.setVisitPlace(visit.getLocation().getDisplay());
        visitEntity.setStartDate(visit.getStartDatetime());
        visitEntity.setStopDate(visit.getStopDatetime());
        return visitEntity;
    }

    public static Patient patientEntityToPatient(PatientEntity patientEntity) {
        Patient patient = new Patient(patientEntity.getId(), patientEntity.getEncounters(), null);
        patient.setDisplay(patientEntity.getDisplay());
        patient.setUuid(patientEntity.getUuid());

        PatientIdentifier patientIdentifier = new PatientIdentifier();
        patientIdentifier.setIdentifier(patientEntity.getIdentifier());
        if (patient.getIdentifiers() == null) {
            patient.setIdentifiers(new ArrayList<>());
        }
        patient.getIdentifiers().add(patientIdentifier);

        PersonName personName = new PersonName();
        personName.setGivenName(patientEntity.getGivenName());
        personName.setMiddleName(patientEntity.getMiddleName());
        personName.setFamilyName(patientEntity.getFamilyName());
        patient.getNames().add(personName);

        patient.setGender(patientEntity.getGender());
        patient.setBirthdate(patientEntity.getBirthDate());
        byte[] photoByteArray = patientEntity.getPhoto();
        if (photoByteArray != null) {
            patient.setPhoto(byteArrayToBitmap(photoByteArray));
        }

        PersonAddress personAddress = new PersonAddress();
        personAddress.setAddress1(patientEntity.getAddress_1());
        personAddress.setAddress2(patientEntity.getAddress_2());
        personAddress.setPostalCode(patientEntity.getPostalCode());
        personAddress.setCountry(patientEntity.getCountry());
        personAddress.setStateProvince(patientEntity.getState());
        personAddress.setCityVillage(patientEntity.getCity());
        patient.getAddresses().add(personAddress);

        if (patientEntity.getCauseOfDeath() != null) {
            patient.setCauseOfDeath(new Resource(ApplicationConstants.EMPTY_STRING, patientEntity.getCauseOfDeath(), new ArrayList<>(), 0));
        }
        if (patientEntity.getDeceased().equals("true")) {
            patient.setDeceased(true);
        } else {
            patient.setDeceased(false);
        }

        return patient;
    }

    public static PatientEntity patientToPatientEntity(Patient patient) {
        PatientEntity patientEntity = new PatientEntity();
        patientEntity.setDisplay(patient.getName().getNameString());
        patientEntity.setUuid(patient.getUuid());
        patientEntity.setSynced(patient.isSynced());

        if (patient.getIdentifier() != null) {
            patientEntity.setIdentifier(patient.getIdentifier().getIdentifier());
        } else {
            patientEntity.setIdentifier(null);
        }

        patientEntity.setGivenName(patient.getName().getGivenName());
        patientEntity.setMiddleName(patient.getName().getMiddleName());
        patientEntity.setFamilyName(patient.getName().getFamilyName());
        patientEntity.setGender(patient.getGender());
        patientEntity.setBirthDate(patient.getBirthdate());
        patientEntity.setDeathDate(null);

        if (null != patient.getCauseOfDeath()) {
            if (patient.getCauseOfDeath().getDisplay() == null) {
                patientEntity.setCauseOfDeath(null);
            } else {
                patientEntity.setCauseOfDeath(patient.getCauseOfDeath().getDisplay());
            }
        } else {
            patientEntity.setCauseOfDeath(null);
        }
        patientEntity.setAge(null);

        if (patient.getPhoto() != null) {
            patientEntity.setPhoto(bitmapToByteArray(patient.getPhoto()));
        } else {
            patientEntity.setPhoto(null);
        }

        if (null != patient.getAddress()) {
            patientEntity.setAddress_1(patient.getAddress().getAddress1());
            patientEntity.setAddress_2(patient.getAddress().getAddress2());
            patientEntity.setPostalCode(patient.getAddress().getPostalCode());
            patientEntity.setCountry(patient.getAddress().getCountry());
            patientEntity.setState(patient.getAddress().getStateProvince());
            patientEntity.setCity(patient.getAddress().getCityVillage());
        }

        patientEntity.setEncounters(patient.getEncounters());
        patientEntity.setDeceased(patient.isDeceased().toString());

        return patientEntity;
    }

    private static byte[] bitmapToByteArray(Bitmap image) {
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        image.compress(Bitmap.CompressFormat.PNG, 0, outputStream);
        return outputStream.toByteArray();
    }

    private static Bitmap byteArrayToBitmap(byte[] imageByteArray) {
        ByteArrayInputStream inputStream = new ByteArrayInputStream(imageByteArray);
        return BitmapFactory.decodeStream(inputStream);
    }

    public static Allergy allergyEntityToAllergy(AllergyEntity allergyEntity) {
        Allergy allergy = new Allergy();
        allergy.setId(allergyEntity.getId());

        allergy.setComment(allergyEntity.getComment());

        if (allergyEntity.getAllergyReactions() != null) {
            allergy.setReactions(allergyEntity.getAllergyReactions());
        } else {
            allergy.setReactions(new ArrayList<>());
        }

        Allergen allergen = new Allergen();
        allergen.setCodedAllergen(new Resource(allergyEntity.getAllergenUUID(), allergyEntity.getAllergenDisplay(), new ArrayList<>(), 1));
        allergy.setAllergen(allergen);

        if (allergyEntity.getSeverityDisplay() != null) {
            allergy.setSeverity(new Resource(allergyEntity.getSeverityUUID(), allergyEntity.getSeverityDisplay(), new ArrayList<>(), 1));
        }
        return allergy;
    }

    public static AllergyEntity allergyToEntity(Allergy allergy, String patientID) {
        AllergyEntity allergyEntity = new AllergyEntity();
        allergyEntity.setPatientId(patientID);
        allergyEntity.setComment(allergy.getComment());
        if (allergy.getSeverity() != null) {
            allergyEntity.setSeverityDisplay(allergy.getSeverity().getDisplay());
            allergyEntity.setSeverityUUID(allergy.getSeverity().getUuid());
        }
        allergyEntity.setAllergenDisplay(allergy.getAllergen().getCodedAllergen().getDisplay());
        allergyEntity.setAllergenUUID(allergy.getAllergen().getCodedAllergen().getUuid());
        allergyEntity.setAllergyReactions(allergy.getReactions());
        return allergyEntity;
    }

    public static List<Allergy> allergyEntityListToAllergyList(List<AllergyEntity> entities) {
        ArrayList<Allergy> allergies = new ArrayList<>();
        for (AllergyEntity allergyEntity : entities) {
            allergies.add(AppDatabaseHelper.allergyEntityToAllergy(allergyEntity));
        }
        return allergies;
    }

    public static <T> Observable<T> createObservableIO(final Callable<T> func) {
        return Observable.fromCallable(func)
                .subscribeOn(Schedulers.io());
    }

}

File: openmrs-android-sdk/src/main/java/com/example/openmrs_android_sdk/library/databases/AppDatabaseHelper.kt
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */
package com.example.openmrs_android_sdk.library.databases

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.example.openmrs_android_sdk.library.OpenmrsAndroid
import com.example.openmrs_android_sdk.library.dao.EncounterDAO
import com.example.openmrs_android_sdk.library.dao.ObservationDAO
import com.example.openmrs_android_sdk.library.dao.PatientDAO
import com.example.openmrs_android_sdk.library.databases.entities.*
import com.example.openmrs_android_sdk.library.models.*
import com.example.openmrs_android_sdk.utilities.ApplicationConstants
import com.example.openmrs_android_sdk.utilities.DateUtils
import com.example.openmrs_android_sdk.utilities.DateUtils.convertTime
import com.example.openmrs_android_sdk.utilities.FormService.getFormByUuid
import rx.Observable
import rx.schedulers.Schedulers
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.util.concurrent.Callable

object AppDatabaseHelper {
    @JvmStatic
    fun convert(obs: Observation, encounterID: Long): ObservationEntity {
        val observationEntity = ObservationEntity()
        observationEntity.id = obs.id
        observationEntity.uuid = obs.uuid
        observationEntity.display = obs.display
        observationEntity.encounterKeyID = encounterID
        observationEntity.displayValue = obs.displayValue
        observationEntity.diagnosisOrder = obs.diagnosisOrder
        observationEntity.diagnosisList = obs.diagnosisList
        observationEntity.diagnosisCertainty = obs.diagnosisCertainty
        observationEntity.diagnosisNote = obs.diagnosisNote
        if (obs.concept != null) {
            observationEntity.conceptuuid = obs.concept!!.uuid
        } else {
            observationEntity.conceptuuid = null
        }
        return observationEntity
    }

    @JvmStatic
    fun convert(observationEntityList: List<ObservationEntity>): List<Observation> {
        val observationList: MutableList<Observation> = ArrayList()
        for (entity in observationEntityList) {
            val obs = Observation()
            obs.id = entity.id
            obs.encounterID = entity.encounterKeyID
            obs.uuid = entity.uuid
            obs.display = entity.display
            obs.displayValue = entity.displayValue
            obs.diagnosisOrder = entity.diagnosisOrder
            obs.diagnosisList = entity.diagnosisList
            obs.setDiagnosisCertanity(entity.diagnosisCertainty)
            obs.diagnosisNote = entity.diagnosisNote
            val concept = ConceptEntity()
            concept.uuid = entity.conceptuuid
            obs.concept = concept
            observationList.add(obs)
        }
        return observationList
    }

    @JvmStatic
    fun convert(encounter: Encounter, visitID: Long?): EncounterEntity {
        val encounterEntity = EncounterEntity()
        encounterEntity.id = encounter.id
        encounterEntity.display = encounter.display
        encounterEntity.uuid = encounter.uuid
        if (visitID != null) {
            encounterEntity.visitKeyId = visitID.toString()
        }
        encounterEntity.encounterDateTime = encounter.encounterDatetime.toString()
        encounterEntity.encounterType = encounter.encounterType!!.display
        encounterEntity.patientUuid = encounter.patientUUID
        encounterEntity.formUuid = encounter.formUuid
        if (null == encounter.location) {
            encounterEntity.locationUuid = null
        } else {
            encounterEntity.locationUuid = encounter.location!!.uuid
        }
        if (encounter.encounterProviders.isEmpty()) {
            encounterEntity.encounterProviderUuid = null
        } else {
            encounterEntity.encounterProviderUuid = encounter.encounterProviders[0].uuid
        }
        return encounterEntity
    }

    @JvmStatic
    fun convert(entity: EncounterEntity): Encounter {
        val encounter = Encounter()
        if (null != entity.encounterType) {
            encounter.encounterType = EncounterType(entity.encounterType)
        }
        encounter.id = entity.id
        if (null != entity.visitKeyId) {
            encounter.visitID = entity.visitKeyId.toLong()
        }
        encounter.uuid = entity.uuid
        encounter.display = entity.display
        val dateTime = entity.encounterDateTime.toLong()
        encounter.setEncounterDatetime(convertTime(dateTime, DateUtils.OPEN_MRS_REQUEST_FORMAT))
        encounter.observations = ObservationDAO().findObservationByEncounterID(entity.id)
        val location: LocationEntity? = try {
            AppDatabase
                    .getDatabase(OpenmrsAndroid.getInstance()?.applicationContext)
                    .locationRoomDAO()
                    .findLocationByUUID(entity.locationUuid)
                    .blockingGet()
        } catch (e: Exception) {
            null
        }
        encounter.location = location
        encounter.form = getFormByUuid(entity.formUuid)
        return encounter
    }

    @JvmStatic
    fun convert(visitEntity: VisitEntity): Visit {
        val visit = Visit()
        visit.id = visitEntity.id
        visit.uuid = visitEntity.uuid
        visit.display = visitEntity.display
        visit.visitType = VisitType(visitEntity.visitType)
        try {
            val locationEntity = AppDatabase
                    .getDatabase(OpenmrsAndroid.getInstance()?.applicationContext)
                    .locationRoomDAO()
                    .findLocationByName(visitEntity.visitPlace)
                    .blockingGet()
            visit.location = locationEntity
        } catch (e: Exception) {
            visit.location = LocationEntity(visitEntity.visitPlace)
        }
        visit.startDatetime = visitEntity.startDate
        visit.stopDatetime = visitEntity.stopDate
        visit.encounters = EncounterDAO().findEncountersByVisitID(visitEntity.id)
        visit.patient = PatientDAO().findPatientByID(visitEntity.patientKeyID.toString())
        return visit
    }

    @JvmStatic
    fun convert(visit: Visit): VisitEntity {
        val visitEntity = VisitEntity()
        visitEntity.id = visit.id
        visitEntity.uuid = visit.uuid
        visitEntity.patientKeyID = visit.patient.id!!
        visitEntity.visitType = visit.visitType.display
        visitEntity.visitPlace = visit.location.display
        visitEntity.isStartDate = visit.startDatetime
        visitEntity.stopDate = visit.stopDatetime
        return visitEntity
    }

    @JvmStatic
    fun convert(patientEntity: PatientEntity): Patient {
        val patient = Patient(patientEntity.id, patientEntity.encounters, null)
        patient.display = patientEntity.display
        patient.uuid = patientEntity.uuid
        val patientIdentifier = PatientIdentifier()
        patientIdentifier.identifier = patientEntity.identifier
        if (patient.identifiers == null) {
            patient.identifiers = ArrayList()
        }
        patient.identifiers.add(patientIdentifier)
        val personName = PersonName()
        personName.givenName = patientEntity.givenName
        personName.middleName = patientEntity.middleName
        personName.familyName = patientEntity.familyName
        patient.names.add(personName)
        patient.gender = patientEntity.gender
        patient.birthdate = patientEntity.birthDate
        val photoByteArray = patientEntity.photo
        if (photoByteArray != null) {
            patient.photo = byteArrayToBitmap(photoByteArray)
        }
        val personAddress = PersonAddress()
        personAddress.address1 = patientEntity.address_1
        personAddress.address2 = patientEntity.address_2
        personAddress.postalCode = patientEntity.postalCode
        personAddress.country = patientEntity.country
        personAddress.stateProvince = patientEntity.state
        personAddress.cityVillage = patientEntity.city
        patient.addresses.add(personAddress)
        if (patientEntity.causeOfDeath != null) {
            patient.causeOfDeath = Resource(ApplicationConstants.EMPTY_STRING, patientEntity.causeOfDeath, ArrayList(), 0)
        }
        patient.isDeceased = patientEntity.deceased == "true"
        return patient
    }

    @JvmStatic
    fun convert(patient: Patient): PatientEntity {
        val patientEntity = PatientEntity()
        patientEntity.display = patient.name.nameString
        patientEntity.uuid = patient.uuid
        patientEntity.isSynced = patient.isSynced
        if (patient.identifier != null) {
            patientEntity.identifier = patient.identifier.identifier
        } else {
            patientEntity.identifier = null
        }
        patientEntity.givenName = patient.name.givenName
        patientEntity.middleName = patient.name.middleName
        patientEntity.familyName = patient.name.familyName
        patientEntity.gender = patient.gender
        patientEntity.birthDate = patient.birthdate
        patientEntity.deathDate = null
        if (null != patient.causeOfDeath) {
            if (patient.causeOfDeath.display == null) {
                patientEntity.causeOfDeath = null
            } else {
                patientEntity.causeOfDeath = patient.causeOfDeath.display
            }
        } else {
            patientEntity.causeOfDeath = null
        }
        patientEntity.age = null
        if (patient.photo != null) {
            patientEntity.photo = bitmapToByteArray(patient.photo)
        } else {
            patientEntity.photo = null
        }
        if (null != patient.address) {
            patientEntity.address_1 = patient.address.address1
            patientEntity.address_2 = patient.address.address2
            patientEntity.postalCode = patient.address.postalCode
            patientEntity.country = patient.address.country
            patientEntity.state = patient.address.stateProvince
            patientEntity.city = patient.address.cityVillage
        }
        patientEntity.encounters = patient.encounters
        patientEntity.deceased = patient.isDeceased.toString()
        return patientEntity
    }

    @JvmStatic
    fun convert(allergy: Allergy, patientID: String?): AllergyEntity {
        val allergyEntity = AllergyEntity()
        allergyEntity.uuid = allergy.uuid
        allergyEntity.patientId = patientID
        allergyEntity.comment = allergy.comment
        if (allergy.severity != null) {
            allergyEntity.severityDisplay = allergy.severity!!.display
            allergyEntity.severityUUID = allergy.severity!!.uuid
        }
        allergyEntity.allergenDisplay = allergy.allergen!!.codedAllergen!!.display
        allergyEntity.allergenUUID = allergy.allergen!!.codedAllergen!!.uuid
        allergyEntity.allergenType = allergy.allergen!!.allergenType
        allergyEntity.allergyReactions = allergy.reactions
        return allergyEntity
    }

    @JvmStatic
    @JvmName("convertTo")
    fun convert(entities: List<AllergyEntity>): List<Allergy> {
        val allergies = ArrayList<Allergy>()
        for (allergyEntity in entities) {
            allergies.add(convert(allergyEntity))
        }
        return allergies
    }

    @JvmStatic
    fun convert(allergyEntity: AllergyEntity): Allergy {
        val allergy = Allergy()
        allergy.id = allergyEntity.id
        allergy.uuid = allergyEntity.uuid
        allergy.comment = allergyEntity.comment
        if (allergyEntity.allergyReactions != null) {
            allergy.reactions = allergyEntity.allergyReactions!!
        } else {
            allergy.reactions = ArrayList()
        }
        val allergen = Allergen()
        allergen.allergenType = allergyEntity.allergenType
        allergen.codedAllergen = Resource(allergyEntity.allergenUUID!!, allergyEntity.allergenDisplay!!, ArrayList(), 1)
        allergy.allergen = allergen
        if (allergyEntity.severityDisplay != null) {
            allergy.severity = Resource(allergyEntity.severityUUID!!, allergyEntity.severityDisplay!!, ArrayList(), 1)
        }
        return allergy
    }

    private fun bitmapToByteArray(image: Bitmap): ByteArray {
        val outputStream = ByteArrayOutputStream()
        image.compress(Bitmap.CompressFormat.PNG, 0, outputStream)
        return outputStream.toByteArray()
    }

    private fun byteArrayToBitmap(imageByteArray: ByteArray): Bitmap {
        val inputStream = ByteArrayInputStream(imageByteArray)
        return BitmapFactory.decodeStream(inputStream)
    }

    @JvmStatic
    fun <T> createObservableIO(func: Callable<T>?): Observable<T> {
        return Observable.fromCallable(func)
                .subscribeOn(Schedulers.io())
    }
}